feat: srw (#1586)
aws/aws-advanced-jdbc-wrapper: docs/using-the-jdbc-driver/using-plugins/UsingTheSimpleReadWriteSplittingPlugin.md

| `srwReadEndpoint`         | String  |    No    | The endpoint to connect to when `setReadOnly(true)` is called.                                | `null`        | `<cluster-name>.cluster-ro-<XYZ>.<region>.rds.amazonaws.com` |
| `verifyNewSrwConnections` | Boolean |    No    | Enables role-verification for new connections made by the simple read/write splitting plugin. | `true`        | `true`, `false`                                              |

### Limitations of `verifyNewSrwConnections`

The property `verifyNewSrwConnections` is enabled by default. This means that when new connections are made with the simple read/write splitting plugin, a query is sent to the new connection to verify its role. If a write connection connects to a reader, an error is thrown. If a read connection connects to the writer, a warning is logged and the connection is not stored for reuse.

#### Non-Aurora clusters
The verification step uses an Aurora specific query to determine the role of the connection, if the endpoint is not to an Aurora database, disable `verifyNewSrwConnections` by setting it to `false` in the properties.

#### Autocommit

> If `verifyNewSrwConnections` is enabled and autocommit is set to false when the simple read/write splitting plugin makes a new connection, unexpected errors such as `Cannot change transaction read-only property in the middle of a transaction.` will occur. As part of verification, the plugin executes a query against a new connection, which when autocommit is set to false will open a transaction. 

Reviewer: Can we rephrase it? It's hard to understand when the occurs and what to do to fix it.

If autocommit is essential to a workflow, either establish both internal reader and writer connections before setting autocommit to false or disable `verifyNewSrwConnections` by setting it to `false` in the properties.

## Supplying the connection string

> When using the simple read/write splitting plugin against Aurora clusters, you do not have to supply multiple instance URLs in the connection string. Instead, supply just the URL for the initial instance to which you're connecting. You must also include either the failover plugin or the Aurora host list plugin in your plugin chain so that the driver knows to query Aurora for its topology. See the section on [loading the simple read/write splitting plugin](#loading-the-simple-readwrite-splitting-plugin) for more info.

Reviewer: What is `Aurora host list plugin` ?

## Using the Simple Read/Write Splitting Plugin with RDS Proxy

RDS Proxy provides connection pooling and management that significantly improves application scalability by reducing database connection overhead and enabling thousands of concurrent connections through
connection multiplexing. Connecting exclusively through the proxy endpoint ensures consistent connection management, automatic failover handling, and centralized monitoring, while protecting the underlying database from connection exhaustion
and providing a stable abstraction layer that remains consistent even when database topology changes. By providing the read/write endpoint and a read-only endpoint to the simple read/write splitting plugin, the AWS JDBC Driver will connect using 
these endpoints any time setReadOnly is called. 

To take full advantage of the benefits of RDS Proxy it is recommended to only connect through RDS Proxy endpoints. See [Using the AWS JDBC Driver with RDS Proxy](./../../../README.md#rds-proxy) for limitations.

> ## Using the Simple Read/Write Splitting Plugin against non-Aurora clusters

Reviewer: non-Aurora and non-RDS Clusters?

> The simple read/write splitting plugin can be used to switch between any two endpoints. If the endpoints do not direct to an Aurora database, ensure the property `verifyNewSrwConnections` is set to `false`. See [Limitations of verifyNewSrwConnections](UsingTheSimpleReadWriteSplittingPlugin.md#non-aurora-clusters) for details.

Reviewer: the same, Aurora or RDS Cluster

## Limitations

### General plugin limitations

When a Statement or ResultSet is created, it is internally bound to the database connection established at that moment. There is no standard JDBC functionality to change the internal connection used by Statement or ResultSet objects. Consequently, even if the read/write plugin switches the internal connection, any Statements/ResultSets created before this will continue using the old database connection. This bypasses the desired functionality provided by the plugin. To prevent these scenarios, an exception will be thrown if your code uses any Statements/ResultSets created before a change in internal connection. To solve this problem, please ensure you create new Statement/ResultSet objects after switching between the writer/reader.

Verify plugin compatibility within your driver configuration using the [compatibility guide](../Compatibility.md).

### Failover

> Immediately following a failover event, due to DNS caching, an Aurora cluster endpoint may connect to the previous writer, and the read-only endpoint may connect to the new primary instance. To avoid stale DNS connections, as well as errors/warnings from the verification logic we recommend including the [Aurora Initial Connection Strategy Plugin](./UsingTheAuroraInitialConnectionStrategyPlugin.md) in your configuration.

Reviewer: I think that using `initialConnection` plugin ruins `srw` plugin. I don't think they're compatible.

```

## Simple Read/Write Splitting Plugin Parameters
| Parameter                    |  Value  | Required | Description                                                                                                                                   | Default Value | Example Value                                                |
|------------------------------|:-------:|:--------:|:----------------------------------------------------------------------------------------------------------------------------------------------|---------------|--------------------------------------------------------------|
| `srwWriteEndpoint`           | String  |   Yes    | The endpoint to connect to when `setReadOnly(false)` is called.                                                                               | `null`        | `<cluster-name>.cluster-<XYZ>.<region>.rds.amazonaws.com`    |
| `srwReadEndpoint`            | String  |   Yes    | The endpoint to connect to when `setReadOnly(true)` is called.                                                                                | `null`        | `<cluster-name>.cluster-ro-<XYZ>.<region>.rds.amazonaws.com` |

> | `verifyNewSrwConnections`    | Boolean |    No    | Enables role verification for new connections made by the Simple Read/Write Splitting Plugin.                                                 | `true`        | `true`, `false`                                              |

Reviewer 2: What is the difference between `verifyNewSrwConnections` and `verifyOpenedConnectionType`? It sounds like the first enables verification of both readers and writers when setReadOnly is called, but the second enables verification for only "writers" or only "readers"? If the first is set but the second is not, does it verify both?

| `srwConnectRetryIntervalMs`  | Integer |    No    | If `verifyNewSrwConnections` is set to `true`, this parameter sets the time delay in milliseconds between each retry of opening a connection. | `1000`        | `1000`                                                       |

## How the Simple Read/Write Splitting Plugin Verifies Connections

The property `verifyNewSrwConnections` is enabled by default. This means that when new connections are made with the Simple Read/Write Splitting Plugin, a query is sent to the new connection to verify its role. If the connection cannot be verified as having the correct role—that is, a write connection is not connected to a writer, or a read connection is not connected to a reader—the plugin will retry the connection up to the time limit of `srwConnectRetryTimeoutMs`. 

The values of `srwConnectRetryTimeoutMs` and `srwConnectRetryIntervalMs` control the timing and aggressiveness of the plugin's retries.

> Additionally, to consistently ensure the role of connections made with the plugin, the plugin also provides role verification for the initial connection. When connecting with an RDS writer cluster or reader cluster endpoint, the plugin will retry the initial connection up to `srwConnectRetryTimeoutMs` until it has verified connection to the intended role of the endpoint.

Reviewer 2: ```suggestion
Additionally, to consistently ensure the role of connections made with the plugin, the plugin also provides role verification for the initial connection. When connecting with an RDS writer cluster or reader cluster endpoint, the plugin will retry the initial connection up to `srwConnectRetryTimeoutMs` until it has verified the intended role of the endpoint.
```

> If it is unable to return a verified initial connection, it will log a message and continue with the normal workflow of the other plugins.

Reviewer 2: Is this referring to the case when we successfully check the role but it does not align with what we expected? Or is this the case when we cannot determine the role we are connected to?

When connecting with custom endpoints and other non-standard URLs, role verification on the initial connection can also be triggered by providing the expected role through the `verifyOpenedConnectionType` parameter. Set this to `writer` or `reader` accordingly.

## Limitations When Verifying Connections

#### Non-RDS clusters

> The verification step determines the role of the connection by executing a query against it. If the endpoint is not part of an Aurora or RDS cluster, disable `verifyNewSrwConnections` by setting it to `false` in the properties.

Reviewer 2: ```suggestion
The verification step determines the role of the connection by executing a query against it. If the endpoint is not part of an Aurora or RDS cluster, the plugin will not be able to verify the role, so `verifyNewSrwConnections` must be set to `false`.
```

#### Autocommit
The verification logic results in errors such as `Cannot change transaction read-only property in the middle of a transaction` from the underlying driver when:
- autocommit is set to false 
- setReadOnly is called 
- as part of setReadOnly, a new connection is opened
- that connection's role is verified

This is a result of the plugin executing the role-verification query against a new connection, and when autocommit is false, this opens a transaction. 

If autocommit is essential to a workflow, either ensure the plugin has connected to the desired target connection of the setReadOnly query before setting autocommit to false or disable `verifyNewSrwConnections`. Examples of the former can be found in the [Simple Read/Write Splitting Examples](UsingTheSimpleReadWriteSplittingPlugin.md#examples).

## Using the Simple Read/Write Splitting Plugin with RDS Proxy

> RDS Proxy provides connection pooling and management that significantly improve application scalability by reducing database connection overhead and enabling thousands of concurrent connections through

Reviewer 2: ```suggestion
RDS Proxy provides connection pooling and management that significantly improves application scalability by reducing database connection overhead and enabling thousands of concurrent connections through
```
